packages/bitcore-wallet-client/lib/key.js

```diff
@@ -0,0 +1,406 @@
+  return x;
+};
+
+
+
+Key.prototype.toObj = function() {
+  var self = this;
+
+  var x = {};
+  _.each(Key.FIELDS, function(k) {
+    x[k] = self[k];
+  });
+  return x;
+};
+
+Key.prototype.isPrivKeyEncrypted = function() {
+  return (!!this.xPrivKeyEncrypted) && !this.xPrivKey;
+};
+
+
+Key.prototype.checkPassword = function(password) {
+  if (this.isPrivKeyEncrypted()) {
+    try {
+      sjcl.decrypt(password, this.xPrivKeyEncrypted);
+    } catch (ex) {
+      return false;
+    }
+    return true;
+  } 
+  return null;
+};
+
+Key.prototype.get = function(password) {
+  var keys = {};
+  let fingerPrintUpdated = false;
+
+  if (this.isPrivKeyEncrypted()) {
+    $.checkArgument(password, 'Private keys are encrypted, a password is needed');
+    try {
+      keys.xPrivKey = sjcl.decrypt(password, this.xPrivKeyEncrypted);
+
+      // update fingerPrint if not set.
+      if (!this.fingerPrint) {
+        let xpriv = new Bitcore.HDPrivateKey(keys.xPrivKey);
+        this.fingerPrint = xpriv.fingerPrint.toString('hex');
+        fingerPrintUpdated = true;
+      }
+
+      if (this.mnemonicEncrypted) {
+        keys.mnemonic = sjcl.decrypt(password, this.mnemonicEncrypted);
+      }
+    } catch (ex) {
+      throw new Error('Could not decrypt');
+    }
+  } else {
+    keys.xPrivKey = this.xPrivKey;
+    keys.mnemonic = this.mnemonic;
+    if (fingerPrintUpdated) {
+      keys.fingerPrintUpdated = true;
+    }
+  }
+  return keys;
+};
+
+Key.prototype.encrypt = function(password, opts) {
+  if (this.xPrivKeyEncrypted)
+    throw new Error('Private key already encrypted');
+
+  if (!this.xPrivKey)
+    throw new Error('No private key to encrypt');
+
+  this.xPrivKeyEncrypted = sjcl.encrypt(password, this.xPrivKey, opts);
+  if (!this.xPrivKeyEncrypted)
+    throw new Error('Could not encrypt');
+
+  if (this.mnemonic)
+    this.mnemonicEncrypted = sjcl.encrypt(password, this.mnemonic, opts);
+
+  delete this.xPrivKey;
+  delete this.mnemonic;
+};
+
+Key.prototype.decrypt = function(password) {
+  if (!this.xPrivKeyEncrypted)
+    throw new Error('Private key is not encrypted');
+
+  try {
+    this.xPrivKey = sjcl.decrypt(password, this.xPrivKeyEncrypted);
+    if (this.mnemonicEncrypted) {
+      this.mnemonic = sjcl.decrypt(password, this.mnemonicEncrypted);
+    }
+    delete this.xPrivKeyEncrypted;
+    delete this.mnemonicEncrypted;
+  } catch (ex) {
+    log.error('error decrypting:', ex);
+    throw new Error('Could not decrypt');
+  }
+};
+
+
+Key.prototype.derive = function(password, path) {
+  $.checkArgument(path, 'no path at derive()');
+  var xPrivKey = new Bitcore.HDPrivateKey(this.get(password).xPrivKey,NETWORK);
+  var deriveFn = this.compliantDerivation ? _.bind(xPrivKey.deriveChild, xPrivKey) : _.bind(xPrivKey.deriveNonCompliantChild, xPrivKey);
+  return deriveFn(path);
+};
+
+function _checkCoin(coin) {
+  if (!_.includes(['btc', 'bch'], coin)) throw new Error('Invalid coin');
```

Comment:
What do you think about making a Coin array in Common.Constants instead of hard coding here?

Reply:
💯 thanks.

Reply:
done.